### PR TITLE
Feature/hard stop for turret

### PR DESCRIPTION
**Pull request by ZilongZeng**
created left and right limits for the turret

### CODE REVIEW

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java [hunk 1]
```diff
@@ -115,6 +119,14 @@ public boolean belowMinimumAngle() {
         return getCurrentAngle() <= minAngleProp.get();
     }
 
+    public boolean LeftLimitStop(){
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/turret/TurretSubsystem.java]: I suggest renaming this to `isAtLeftHardStop()`

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java [hunk 2]
```diff
@@ -115,6 +119,14 @@ public boolean belowMinimumAngle() {
...
+        return motor.isRevLimitSwitchClosed();
+    }
+
+    public boolean rightLimitStop(){
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/turret/TurretSubsystem.java]: see above

**Review by JohnGilb — CHANGES_REQUESTED** [review]
where are your tests?

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java (hunk 2 above)
Comment (aschokking) [reply src/main/java/competition/subsystems/turret/TurretSubsystem.java]: would be more clear if you made this `isAtLeftLimitStop` as John suggested earlier

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java [hunk 3]
```diff
@@ -86,17 +90,17 @@ public void turnLeft() {
...
             // No sense running the protection code if we don't know where we are.
             
             // Check for any reason power should be constrained.
-            if (aboveMaximumAngle()) {
+            if (aboveMaximumAngle() || rightLimitStop()) {
                 // Turned too far left. Only allow right/negative rotation.
                 power = MathUtils.constrainDouble(power, -1, 0);
             }
-            if (belowMinimumAngle()) {
+            if (belowMinimumAngle() || leftLimitStop()) {
```
Comment (aschokking) [inline src/main/java/competition/subsystems/turret/TurretSubsystem.java]: consider renaming yours or belowMinimumAngle so they both talk about left/right or min/max angle but not a mix of them

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java [hunk 4]
```diff
@@ -101,12 +105,23 @@ public void setPower(double power) {
                 power = MathUtils.constrainDouble(power, 0, 1);
             }
         }
+        if (isAtMaxHardStop()) {
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/turret/TurretSubsystem.java:114]: we still need the checks for min/max angle. 
if (`isAtMaxHardStop() || aboveMaximumAngle())`

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java [hunk 5]
```diff
@@ -101,12 +105,23 @@ public void setPower(double power) {
...
+            // Turned too far left. Only allow right/negative rotation.
+            power = MathUtils.constrainDouble(power, -1, 0);
+        }
+        if (isAtMinHardStop()) {
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/turret/TurretSubsystem.java:118]: see above

**Review by JohnGilb — CHANGES_REQUESTED** [review]
we need to check for both limits and "soft" angles when stopping the motor.

File: src/main/java/competition/subsystems/turret/TurretSubsystem.java (hunk 4 above)
Comment (aschokking) [reply src/main/java/competition/subsystems/turret/TurretSubsystem.java:114]: @JohnGilb it is still checking the angle's above, but it only happens if the turret is calibrated (whereas these hardstop checks should happen regardless) so I think the current setup is correct?

https://github.com/Team488/TeamXbot2020/pull/39/files/a3fed3aa4e6d8698901045f1aa1fb75496bec70a#diff-db9af9f03585fe2c810b92d48732a930L95

**Review by JohnGilb — APPROVED** [review]

**Pull request merged**